content/courses/advanced-ada/parts/resource_management/controlled_types.rst

```diff
@@ -3,14 +3,551 @@ Controlled Types
+:wikipedia:`lifetime of objects <Variable_(computer_science)#Scope_and_extent>`
+previously in the context of
+:ref:`access types <Adv_Ada_Access_Types_Lifetime_of_Objects>`.
+Again, we assume you understand the concept. In any case, let's quickly review
+the typical lifetime of an object:
+
+.. uml::
+   :align: center
+   :width: 100pt
+
+    @startuml
+    start
+    :Create object A;
+    :Use object A;
+    :Finalize object A;
+    stop
+    @enduml
+
+In simple terms, an object :ada:`A` is first created before we can make use of
+it. When object :ada:`A` is about to get out of scope, it is finalized.
+
+Let's analyze the lifetime of object :ada:`A` in a procedure :ada:`P`:
+
+.. code-block:: ada
+
+    procedure P is
+       A : T;
+    begin
+       P2 (A);
+    end P;
+
+We could visualize the lifetime as follows:
+
+.. uml::
+   :align: center
+   :width: 230pt
+
+    @startuml
+    start
+    group block's declarative part
+        :Create object A;
+    end group
+    group block's handled sequence of statements;
+        :Use object A;
+    end group
+    group block's end part
+        :Destroy object A;
+    end group
+    stop
+    @enduml
+
+In other words, object :ada:`A` is created in the declarative part of :ada:`P`
+and then it's used in :ada:`P`\'s sequence of statements. Finally, :ada:`A` is
+finalized when :ada:`P` ends.
+
+
+Initialization of objects
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Typically, right after an object :ada:`A` is created, it is still uninitialized.
+Therefore, we have to explicitly initialize it with a meaningful initial value
+|mdash| or with the value returned by a function call, for example. Similarly,
+when an object :ada:`A` is about to get out of scope, it is going to be
+finalized (i.e. destroyed) and its contents are then lost forever.
+
+As we know, for some standard Ada types, objects are initialized by default.
+For example, objects of access types are initialized by default to :ada:`null`.
+Likewise, we can declare
+:ref:`types with default initial value <Adv_Ada_Default_Initial_Values>`:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Controlled_Types.Overview.Default_Initialization
+
+    pragma Ada_2022;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Main is
+       type Int is new Integer
+         with Default_Value => 42;
+
+       I  : Int;
+       AI : access Int;
+    begin
+       Put_Line ("I : "
+                 & I'Image);
+       Put_Line ("AI : "
+                 & AI'Image);
+    end Main;
+
+In this case, we can visualize the lifetime of those objects as follows:
+
+.. uml::
+   :align: center
+   :width: 100pt
+
+    @startuml
+    start
+    :Create object A;
+    :Initialize object A
+     with default value;
+    :Use object A;
+    :Finalize object A;
+    stop
+    @enduml
+
+Even though these default initialization methods provide some control over the
+objects, they're not very flexible. Also, we don't have any means to perform
```

Comment:
I think that's going too far. Record types have a very good default initialization capability IMO. And they are the most common completion for private types so the facility is often used. I'd say default initialization is the first choice since it is guaranteed and requires nothing of the client. And is cheap at run-time compared to controlled types.

Reply:
OK, I changed the phrase so it doesn't sound so radical. Also, I added your comments to a small admonition. Please let me know what you think!